binary_transparency/firmware/internal/ftmap/aggregate.go

```diff
@@ -0,0 +1,95 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ftmap
+
+import (
+	"crypto/sha512"
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	"github.com/apache/beam/sdks/go/pkg/beam"
+
+	"github.com/google/trillian-examples/binary_transparency/firmware/api"
+	"github.com/google/trillian/experimental/batchmap"
+	"github.com/google/trillian/merkle/coniks"
+	"github.com/google/trillian/storage/tree"
+)
+
+func init() {
+	beam.RegisterFunction(aggregationFn)
+	beam.RegisterType(reflect.TypeOf((*api.AggregatedFirmware)(nil)).Elem())
+	beam.RegisterType(reflect.TypeOf((*aggregatedFirmwareHashFn)(nil)).Elem())
+}
+
+// Aggregate will output an entry for each firmware entry in the input log, which includes the
+// firmware metadata, along with an aggregated representation of the annotations for it. The
+// rules are:
+// * AnnotationMalware: `Good` is true providing there are no malware annotations that claim the
+//                      firmware is bad.
+func Aggregate(s beam.Scope, treeID int64, fws, annotationMalwares beam.PCollection) (beam.PCollection, beam.PCollection) {
+	keyedFws := beam.ParDo(s, func(l *firmwareLogEntry) (uint64, *firmwareLogEntry) { return uint64(l.Index), l }, fws)
+	keyedAnns := beam.ParDo(s, func(a *annotationMalwareLogEntry) (uint64, *annotationMalwareLogEntry) {
+		return a.Annotation.FirmwareID.LogIndex, a
+	}, annotationMalwares)
+	annotations := beam.ParDo(s, aggregationFn, beam.CoGroupByKey(s, keyedFws, keyedAnns))
+	return beam.ParDo(s, &aggregatedFirmwareHashFn{treeID}, annotations), annotations
+}
+
+func aggregationFn(fwIndex uint64, fwit func(**firmwareLogEntry) bool, amit func(**annotationMalwareLogEntry) bool) (*api.AggregatedFirmware, error) {
+	// There will be exactly one firmware entry for the log index.
+	var fwle *firmwareLogEntry
+	if !fwit(&fwle) {
+		return nil, fmt.Errorf("aggregationFn for %d found no firmware", fwIndex)
+	}
+
+	// And 0-many annotations. The FW is good as long as no annotations say that it is not.
+	good := true
+	var amle *annotationMalwareLogEntry
+	for amit(&amle) {
+		good = good && amle.Annotation.Good
+	}
+
+	return &api.AggregatedFirmware{
+		Index:    fwIndex,
+		Firmware: &fwle.Firmware,
+		Good:     good,
+	}, nil
+}
+
+type aggregatedFirmwareHashFn struct {
+	TreeID int64
+}
+
+func (fn *aggregatedFirmwareHashFn) ProcessElement(afw *api.AggregatedFirmware) (*batchmap.Entry, error) {
+	// We use the Index of the FW Log Entry as the key because that's unique and what we matched on.
```

Comment:
Just curious to know, what would happen in a use case if a log tree is forked and same index is generated twice on different branches, will such a case be detected trying to add the same key ?

Reply:
This map is constructed by consuming the log, so we'll only see one side of the branch. The split-view should be protected against (in a full production map) by:
 * storing the Log Checkpoint & log size the map revision was constructed from
 * checking consistency with the stored log checkpoint before updating the map
 * returning the log checkpoint to map clients so they can check they are on the same log view as the map